File: www/ai/java/src/com/zte/alphattt/game/Board.java
package com.zte.alphattt.game;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Board {
	public static final int PlAYER_1_WIN = 1;
	public static final int PLAYER_2_WIN = 2;
	public static final int DRAW = 3;
	public static final int ON_GOING = 4;
	
	/*
	 * 1 2 4 8 16 32 64 128 256
	 */
	private int[] gameState = new int[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, -1, 1 };
	private int[] posValue = new int[] { 1, 2, 4, 8, 16, 32, 64, 128, 256 };
	private int[] wins = new int[] { 1 + 2 + 4, 8 + 16 + 32, 64 + 128 + 256, 1 + 8 + 64, 2 + 16 + 128, 4 + 32 + 256,
			1 + 16 + 256, 4 + 16 + 64 };
	private static Map<Integer, Integer> lineIndex2grid = new HashMap<Integer, Integer>();
	static {
		lineIndex2grid.put(0, 2);
		lineIndex2grid.put(1, 4);
		lineIndex2grid.put(2, 6);
		lineIndex2grid.put(3, 9);
		lineIndex2grid.put(4, 11);
		lineIndex2grid.put(5, 13);
		lineIndex2grid.put(6, 16);
		lineIndex2grid.put(7, 18);
		lineIndex2grid.put(8, 20);
	}

	public int[] start() {
		return gameState.clone();
	}

	public int[] next_state(int[] state, int[] move) {
		int[] res = state.clone(); 
		if (is_legal(res, move)) {
			int player = res[22];
			int gridIndex = (move[0] * 3 + move[1]) * 2 - 1 + player;
			int gridValue = posValue[move[2] * 3 + move[3]];
			res[gridIndex] = res[gridIndex] + gridValue;
			res[18] = judgeBigWin(1, res);
			res[19] = judgeBigWin(2, res);
			res[20] = move[2];
			res[21] = move[3];
			res[22] = 3 - player;
			return res;
		}
		return null;
	}

	private int judgeBigWin(int player, int[] state) {
		int res = 0;
		for(int i = 0; i < 9; i++){
			if(isWin(state[i*2 + player - 1])){
				res += this.posValue[i];
			}
		}
		return res;
	}

	public int currentPlayer(int[] state){
		return state[22];
	}
	
	public int[][] legal_moves(int[] state) {
		int bGridX = state[20];
		int bGridY = state[21];
		if (bGridX == -1 || bGridY == -1) {
			return findAllLegal(state);
		}

		int gridIndex = (bGridX * 3 + bGridY) * 2 - 1;
		int player1GirdValue = state[gridIndex + 1];
		int player2GridValue = state[gridIndex + 2];
		if (isWin(player1GirdValue) || isWin(player2GridValue) || isFull(player1GirdValue + player2GridValue)) {
			return findAllLegal(state);
		}

		return findGirdLegalMoves(state[gridIndex + 1], state[gridIndex + 2], bGridX, bGridY).toArray(new int[0][]);
	}

	public int max_moves(){
		return 81;
	}
	
	/**
	 * 
	 * @param state
	 * @return 1|2|3:draw|4:ongoing
	 */
	public int winner(int[] state){
		if(isWin(state[18])){
			return PlAYER_1_WIN;
		}
		if(isWin(state[19])){
			return PLAYER_2_WIN;
		}
		if(isFull(state)){
			return DRAW;
		}
		return ON_GOING;
	}
	
	private boolean isFull(int[] state){
		for(int i = 0; i < 9; i++){
			int total = state[i*2] + state[i*2 + 1];
			if(!isFull(total) && !isWin(state[i*2]) && !isWin(state[i*2 + 1])){
				return false;
			}
		}
		return true;
	}
	
	private boolean isFull(int i) {
		return i == 511;
	}

	private List<int[]> findGirdLegalMoves(int player1value, int player2value, int gridx, int gridy) {
		List<int[]> res = new ArrayList<int[]>();
		int total = player1value + player2value;
		for (int i = 0; i < this.posValue.length; i++) {
			if ((total & posValue[i]) == 0) {
				res.add(new int[] { gridx, gridy, i / 3, i % 3 });
			}
		}
		return res;
	}

	private int[][] findAllLegal(int[] state) {
		List<int[]> res = new ArrayList<int[]>();
		for(int i = 0; i < 9; i++){
			int player1GirdValue = state[i*2];
			int player2GridValue = state[i*2 + 1];
			if (isWin(player1GirdValue) || isWin(player2GridValue) || isFull(player1GirdValue + player2GridValue)) {
				continue;
			}
			List<int[]> tmp = findGirdLegalMoves(player1GirdValue, player2GridValue, i / 3, i % 3);
			res.addAll(tmp);
		}
		return res.toArray(new int[0][]);
	}

	private boolean isWin(int i) {
		for (int win : this.wins) {
			if ((i & win) == win) {
				return true;
			}
		}
		return false;
	}

	private boolean is_legal(int[] state, int[] move) {
		return true;
	}

	public void display(int[] state) {
		int[][] wholeGrid = new int[9][9];
		for (int i = 0; i < 9; i++) {
			int[][] grid = parseGrid(state[2 * i], state[2 * i + 1]);
			copyGrid(wholeGrid, grid, i);
		}

		for (int line = 0; line < wholeGrid.length; line++) {
			if (line % 3 == 0) {
				displayBorder();
			}
			this.displayOneLine(wholeGrid[line]);
		}
		displayBorder();
	}

	private void copyGrid(int[][] wholeGrid, int[][] grid, int i) {
		int offsetX = (i / 3) * 3;
		int offsetY = (i % 3) * 3;
		for (int xi = 0; xi < 3; xi++) {
			for (int yi = 0; yi < 3; yi++) {
				wholeGrid[xi + offsetX][yi + offsetY] = grid[xi][yi];
			}
		}
	}

	private int[][] parseGrid(int player1, int player2) {
		int[][] res = new int[][] { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } };
		for (int i = posValue.length - 1; i >= 0; i--) {
			player1 = matchStep(player1, posValue[i], i, 1, res);
			player2 = matchStep(player2, posValue[i], i, 2, res);
		}
		return res;
	}

	private int matchStep(int player, int value, int pos, int desc, int[][] res) {
		if (player / value > 0) {
			player = player - value;
			res[pos / 3][pos % 3] = desc;
		}
		return player;
	}

	private void displayOneLine(int[] line) {
		byte[] bytes = "|| | | || | | || | | ||".getBytes();
		for (int i = 0; i < line.length; i++) {
			if (line[i] == 1) {
				bytes[lineIndex2grid.get(i)] = 'x';
			} else if (line[i] == 2) {
				bytes[lineIndex2grid.get(i)] = 'o';
			}
		}
		System.out.println(new String(bytes));
	}

	public void displayBorder() {
		System.out.println("||-|-|-||-|-|-||-|-|-||");
	}
	
}

File: src/python/client/board.py
# -*- coding: UTF-8 -*-


class Board(object):
    PlAYER_1_WIN = 1
    PLAYER_2_WIN = 2
    DRAW = 3
    ON_GOING = 4
    
    game_state = [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, None, None, 1 ]
    wins = [1 + 2 + 4, 8 + 16 + 32, 64 + 128 + 256, 1 + 8 + 64, 2 + 16 + 128, 4 + 32 + 256,
            1 + 16 + 256, 4 + 16 + 64]
    player1_pos = {(0, 0) : 0,
                       (0, 1) : 2,
                       (0, 2) : 4,
                       (1, 0) : 6,
                       (1, 1) : 8,
                       (1, 2) : 10,
                       (2, 0) : 12,
                       (2, 1) : 14,
                       (2, 2) : 16}
    player2_pos = {(0, 0) : 1,
                       (0, 1) : 3,
                       (0, 2) : 5,
                       (1, 0) : 7,
                       (1, 1) : 9,
                       (1, 2) : 11,
                       (2, 0) : 13,
                       (2, 1) : 15,
                       (2, 2) : 17}
    pos_value_list = [1, 2, 4, 8, 16, 32, 64, 128, 256]
    pos_value = {(0, 0) : 1,
                 (0, 1) : 2,
                 (0, 2) : 4,
                 (1, 0) : 8,
                 (1, 1) : 16,
                 (1, 2) : 32,
                 (2, 0) : 64,
                 (2, 1) : 128,
                 (2, 2) : 256}
    lineIndex2grid = {0 : 2, 1 : 4, 2 : 6, 3 : 9, 4 : 11, 5 : 13, 6 : 16, 7 : 18, 8 : 20}
    index2pos = {0: (0, 0),
                 1: (0, 1),
                 2: (0, 2),
                 3: (1, 0),
                 4: (1, 1),
                 5: (1, 2),
                 6: (2, 0),
                 7: (2, 1),
                 8: (2, 2)}
    def start(self): 
        return list(self.game_state)

    def next_state(self, state, (lr, lc, r, c)):
        res = list(state) 
        player = res[22]
        gridIndex = self.player1_pos[(lr, lc)] if player == 1 else self.player2_pos[(lr, lc)]
        gridValue = self.pos_value[(r, c)]
        res[gridIndex] = res[gridIndex] | gridValue
        res[18] = self.judgeBigWin(1, res, lr, lc)
        res[19] = self.judgeBigWin(2, res, lr, lc)
        res[20] = r
        res[21] = c
        res[22] = 2 if player == 1 else 1
        return res
    
    def judgeBigWin(self, player, state, lr, lc):
        res = 0
        if player == 1:
            if(self.isWin(state[self.player1_pos[(lr, lc)]])):
                res = res | self.pos_value[(lr, lc)]
        if player == 2:
            if(self.isWin(state[self.player2_pos[(lr, lc)]])):
                res = res | self.pos_value[(lr, lc)]
        return res

    def currentPlayer(self, state):
        return state[22]
    
    def legal_moves(self, state):
        if (type(state[20]) is not int or type(state[21]) is not int):
            return self.findAllLegal(state)

        player1GridIndex = self.player1_pos[(state[20], state[21])]
        player2GridIndex = self.player2_pos[(state[20], state[21])]
        player1GirdValue = state[player1GridIndex]
        player2GridValue = state[player2GridIndex]
        if (self.isWin(player1GirdValue) or self.isWin(player2GridValue) or self.isGridFull(player1GirdValue | player2GridValue)) :
            return self.findAllLegal(state)

        return self.findGirdLegalMoves(state[player1GridIndex], state[player2GridIndex], state[20], state[21])

    def max_moves(self):
        return 81;
    
    def winner(self, state):
        if(self.isWin(state[18])):
            return self.PlAYER_1_WIN
        
        if(self.isWin(state[19])):
            return self.PLAYER_2_WIN
        
        if(self.isFull(state)):
            return self.DRAW
        
        return self.ON_GOING
    
    def isFull(self, state):
        for i in range(0, 9):
            total = state[i * 2] + state[i * 2 + 1];
            if((not self.isGridFull(total)) and (not self.isWin(state[i * 2])) and (not self.isWin(state[i * 2 + 1]))):
                return False;
        return True;
    
    def isGridFull(self, i):
        return i == 511;

    def findGirdLegalMoves(self, player1value, player2value, gridx, gridy):
        res = []
        total = player1value | player2value
        for r, c in self.pos_value.keys():
            if total & self.pos_value[(r, c)] == 0 :
                res.append([gridx, gridy, r, c])
        return res;

    def findAllLegal(self, state):
        res = []
        for i in range(0, 9):
            player1GirdValue = state[i*2]
            player2GridValue = state[i*2 + 1]
            if (self.isWin(player1GirdValue) or self.isWin(player2GridValue) or self.isGridFull(player1GirdValue | player2GridValue)):
                continue
            r, c = self.index2pos[i]
            res +=self.findGirdLegalMoves(player1GirdValue, player2GridValue, r, c);
        return res

    def isWin(self, i):
        for win in self.wins:
            if ((i & win) == win):
                return True;
        return False
    
    def display(self, state):
        wholeGrid = [[0 for x in range(0, 9)] for x in range(0, 9)]
        for i in range(0, 9):
            grid = self.parseGrid(state[2 * i], state[2 * i + 1]);
            self.copyGrid(wholeGrid, grid, i);

        for line in range(0, len(wholeGrid)):
            if (line % 3 == 0):
                self.displayBorder();
            self.displayOneLine(wholeGrid[line]);
        self.displayBorder();

    def copyGrid(self, wholeGrid, grid, i):
        offsetX = (i / 3) * 3;
        offsetY = (i % 3) * 3;
        for xi in range(0, 3):
            for yi in range(0, 3):
                wholeGrid[xi + offsetX][yi + offsetY] = grid[xi][yi]

    def parseGrid(self, player1, player2):
        res = [[0, 0, 0 ], [ 0, 0, 0 ], [ 0, 0, 0 ]]
        for i in range(len(self.pos_value_list) - 1, -1, -1):
            player1 = self.matchStep(player1, self.pos_value_list[i], i, 1, res);
            player2 = self.matchStep(player2, self.pos_value_list[i], i, 2, res);
        return res

    def matchStep(self, player, value, pos, desc, res):
        if (player / value > 0) :
            player = player - value;
            res[pos / 3][pos % 3] = desc;
        return player;

    def displayOneLine(self, line):
        s = list("||     ||     ||     ||")
        for i in range(0, len(line)):
            if (line[i] == 1):
                s[self.lineIndex2grid.get(i)] = 'x';
            elif (line[i] == 2):
                s[self.lineIndex2grid.get(i)] = 'o';
        print "".join(s)

    def displayBorder(self):
        print("----------------------")
